extends Node

var player: AudioStreamPlayer
var stream: AudioStream

var playing: bool = false
var elapsed: float = 0
var last_update: int = 0
var playback_speed: float = 1

signal ended

func _ready() -> void:
	player = AudioStreamPlayer.new()
	player.bus=&"Music"
	player.pitch_scale=playback_speed
	player.volume_linear=0.5
	player.finished.connect(func() -> void:
		playing=false
		ended.emit()
	)
	self.add_child(player)

func set_stream(new_stream: AudioStream) -> void:
	stream = new_stream
	player.stream=stream

func set_playback_speed(speed: float) -> void:
	playback_speed=speed
	player.pitch_scale=speed

func play(from:float) -> void:
	playing = true
	player.play()
	elapsed=from
	last_update = Time.get_ticks_usec()

func stop() -> void:
	playing = false
	player.stop()

func resume() -> void:
	play(elapsed)

func _process(_dt: float) -> void:
	if not playing: return
	var current_update: int = Time.get_ticks_usec()
	elapsed += (float(current_update-last_update)/1_000_000)*playback_speed
	last_update = current_update
